textdistance/algorithms/edit_based.py

```diff
@@ -194,11 +200,54 @@ def _numpy(self, s1: Sequence[T], s2: Sequence[T]) -> int:
 
         return d[len(s1) - 1][len(s2) - 1]
 
-    def _pure_python(self, s1: Sequence[T], s2: Sequence[T]) -> int:
+    def _pure_python_unrestricted(self, s1: Sequence[T], s2: Sequence[T]) -> int:
+        """https://en.wikipedia.org/wiki/Damerau%E2%80%93Levenshtein_distance
+        """
+        d: dict[tuple[int, int], int] = {}
+        da: defaultdict[T, int] = defaultdict(int)
+
+        len1 = len(s1)
+        len2 = len(s2)
+
+        maxdist = len1 + len2
+        d[-1, -1] = maxdist
+
+        # matrix
+        for i in range(len(s1) + 1):
+            d[i, -1] = maxdist
+            d[i, 0] = i
+        for j in range(len(s2) + 1):
+            d[-1, j] = maxdist
+            d[0, j] = j
+
+        for i, cs1 in enumerate(s1):
+            i += 1
+            db = 0
+            for j, cs2 in enumerate(s2):
+                j += 1
+                i1 = da[cs2]
+                j1 = db
+                if self.test_func(cs1, cs2):
+                    cost = 0
+                    db = j
+                else:
+                    cost = 1
+
+                d[i, j] = min(
+                    d[i - 1, j - 1] + cost,     # substitution
+                    d[i, j - 1] + 1,            # insertion
+                    d[i - 1, j] + 1,            # deletion
+                    d[i1 - 1, j1 - 1] + (i - i1) - 1 + (j - j1),  # transposition
+                )
+            da[cs1] = i
+
+        return d[len1, len2]
```

Comment:
is this really correct? we write into `d[i, j]` which should be at most `d[len1-1, len2-1]` from my understanding.

Reply:
I didn't modify the code you provided, just moved it around. I don't remember much about the algorithms, I can only rely on tests.

Reply:
I checked the code. What a mess it is, oh my. We first set these values by doing iteration over `range(len(s1) + 1)` when initializing the matrix, And then when we do enumeration `for i, cs1 in enumerate(s1):`, we on the next line shift the index: `i += 1` 🧠

Reply:
This is deliberate!  The boundary values are needed, but are never modified later, as the algorithm looks at the preceding values in the array when calculating the new values.  I tried to come up with a way of avoiding the i+=1 business, but I kept messing things up, so I decided to be somewhat non-Pythonic and stick to the algorithm as presented on Wikipedia!

Reply:
> I tried to come up with a way of avoiding the i+=1 business, but I kept messing things up

Tip of the day: `enumerate` accepts an argument `start`, where you can specify the initial value. I've changed the code to use it. It became a bit slower, for some reason, but IMHO a bit more readable.

Reply:
Yes I completely missed the += 1 before

Reply:
> > I tried to come up with a way of avoiding the i+=1 business, but I kept messing things up
> 
> Tip of the day: `enumerate` accepts an argument `start`, where you can specify the initial value. I've changed the code to use it. It became a bit slower, for some reason, but IMHO a bit more readable.

Oh, how neat! Thanks for that tip. I agree, more readable is better.